### PR TITLE
Update slack workflow canvas action

### PR DESCRIPTION
**Pull request by tomthorogood**
This change list fixes several issues with the overall structure of the "Update Slack Workflow Canvas" action.

The only difficult change to grok is the update to the release workflow itself (the most recent commit), which is better understood via the screenshot below. The other changes (in their own commits) are pretty small and self-documenting. 

The result is a more stable action with a more resilient release workflow.

## Release workflow changes

- Pushes to [GHCR] instead of GCR
- Simplifies to only use one tag stream for now.
- Use org secrets where available (instead of duplicating them in each repo)
- Add steps to show when the workflow is preparing to release, and when it is validating the prepared release (seen below)
- Add branch information to canvas footer (seen below)
- Ensure workflow is always finalized (seen below)

I tested this change by pushing to the `review` branch which generated the following complete workflow in a testing slack channel. 

Screen Shot 2021-04-30 at 3 35 59 AM

[GHCR]: https://ghcr.io/uwit-iam

### CODE REVIEW

File: .github/workflows/update-slack-workflow-canvas-action.yml [hunk 1]
```diff
@@ -10,15 +10,19 @@ on:
       # Any branch beginning with update-slack-workflow-canvas will automatically
       # trigger the testing workflow
       - update-slack-workflow-canvas*
-      - develop  # This branch is used to test the release part of the workflow
-      - main  # This branch actually tags a new release image
+      # When pushed to 'release', the action is image is built and tagged for release,
```
Comment (krlowe) [inline .github/workflows/update-slack-workflow-canvas-action.yml]: This sentence Engrish being reading hard.

File: update-slack-workflow-canvas/action/entrypoint.sh [hunk 2]
```diff
@@ -11,12 +13,33 @@ ACTION_STEP_STATUS="${ACTION_STEP_STATUS}"
...
+# same thing over and over. This will allow us to set
+# values as "use_env(foo)" which will extract the env
+# value at _runtime_, as opposed to during the action's
+# "compile" time. This does so in a way that is safe to use
+# and does not execute arbitrary code using eval.
+# Example:
+#   CANVAS_ID=12345 ACTION_CANVAS="use_env(CANVAS_ID)" ./entrypoint.sh
+function get_value_from_arg() {
+  provided_value="$@"
+  if [[ -n $(echo $provided_value | grep '^use_env') ]]
```
Comment (krlowe) [inline update-slack-workflow-canvas/action/entrypoint.sh]: I think you can replace this pipeline with bash's regex comparison:

    if [[ $provided_value =~ ^use_env ]]

I was playing with that just yesterday ... it took me awhile to figure out you don't want to quote the regex.

File: update-slack-workflow-canvas/action/entrypoint.sh [hunk 3]
```diff
@@ -11,12 +13,33 @@ ACTION_STEP_STATUS="${ACTION_STEP_STATUS}"
...
+  then
+    env_var=$(echo $provided_value | sed 's/^use_env(\(.*\))$/\1/')
+    echo "${!env_var}"
```
Comment (krlowe) [inline update-slack-workflow-canvas/action/entrypoint.sh]: For that matter, let's eliminate this pipeline, too:
```
   if [[ $provided_value =~ ^use_env\((.*)\)$ ]]
   then
       echo "${BASH_REMATCH[1]}"
   else
       echo "$provided_value"
   fi
```
 Or if you want to be totally cryptic:
     `[[ $provided_value =~ ^use_env\((.*)\)$ ]] && echo "${BASH_REMATCH[1]}" || echo "$provided_value"`

**Review by krlowe — APPROVED** [review]
This looks like fun, but is it going to get "busy" ... does the slack message really need to list all things that worked?  E.g., Will we be pulling docker image names from the slack message to paste elsewhere or something?  Or would it be sufficient to have the slack message just say "It worked" or "This step failed" and include a link to the actions report in GitHub for more details?

**Comment by tomthorogood** [comment]
@krlowe It is up to the workflow implementer how busy they would like it to be.

I purposely made the implementation in this particular workflow rather obtuse so that I could test the various use cases the only way I could: by actually running them using github actions.

In "real life," you could absolutely have a single-step workflow, or even something with no steps, and just artifacts attached. The status in the header would still be updated, so it's very flexible :) 

Also, I am going to make a template action for the basic plumbing, so that folks can easily drop it into their repo with the boilerplate filled out already.

**Comment by tomthorogood** [comment]
btw i am considering some of these changes; in the meantime i've pushed this to the `release` branch so that i'm not blocked on my "real" work and can consume this action.

File: update-slack-workflow-canvas/action/entrypoint.sh (hunk 3 above)
Comment (tomthorogood) [reply update-slack-workflow-canvas/action/entrypoint.sh]: This worked after I changed it to `${!BASH_REMATCH[1]}`, thanks! Much cleaner.

**Pull request merged**